### PR TITLE
[Sprint 1]Feature logic choose domain_next

### PR DESCRIPTION
**Pull request by nguyenhuy95dn**

### CODE REVIEW

File: app/src/main/java/com/framgia/forder/data/source/DomainRepository.java [hunk 1]
```diff
@@ -16,4 +19,11 @@ public DomainRepository(DomainRemoteDataSource remoteDataSource,
         mDomainRemoteDataSource = remoteDataSource;
         mDomainLocalDataSource = localDataSource;
     }
+
+    public Observable<List<Domain>> getListDomain() {
+        int user_id = 1;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/DomainRepository.java]: add ``// TODO change to get data from localDataSource``in here!

File: app/src/main/java/com/framgia/forder/data/source/local/DomainLocalDataSource.java [hunk 2]
```diff
@@ -7,7 +7,6 @@
  */
 
 public class DomainLocalDataSource implements DomainDataSource.LocalDataSource {
-    private static final int USER_ID = 1;
-    private static final String USER_EMAIL = "tran.duc.quoc@framgia.com";
-    private static final String USER_TOKEN = "zRow13TqHnUK26teEP7y";
+    public DomainLocalDataSource() {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/DomainLocalDataSource.java]: Need inject ``RealmAPI`` & ``SharedPrefsApi``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 3]
```diff
@@ -1,14 +1,44 @@
...
-public class ChooseDomainViewModel implements ChooseDomainContract.ViewModel {
+public class ChooseDomainViewModel extends BaseObservable
+        implements ChooseDomainContract.ViewModel {
+
+    private static final String EXTRA_ID_DOMAIN = "EXTRA_ID_DOMAIN";
 
+    private Context mContext;
     private ChooseDomainContract.Presenter mPresenter;
+    private ArrayAdapter<String> mAdapter;
+    private List<String> mDomains = new ArrayList<>();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: variable must init in ``constructor`` method

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 4]
```diff
@@ -1,14 +1,44 @@
...
+    private ArrayAdapter<Domain> mAdapterDomain;
+    private List<Domain> mDomainList = new ArrayList<>();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: variable must init in ``constructor`` method

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 5]
```diff
@@ -1,14 +1,44 @@
...
+    private Integer selectedTypePosition;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: check Name convention

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 6]
```diff
@@ -25,4 +55,44 @@ public void onStop() {
     public void setPresenter(ChooseDomainContract.Presenter presenter) {
         mPresenter = presenter;
     }
+
+    @Override
+    public void showListDomain(List<Domain> domains) {
+        updateData(domains);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: delete method ``updateData`` and put all code inside ``updateData`` method into here

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 7]
```diff
@@ -25,4 +55,44 @@ public void onStop() {
...
+    @Override
+    public void onGetDomainError() {
+        Toast.makeText(mContext, R.string.cannot_load_the_data, Toast.LENGTH_SHORT).show();
+    }
+
+    public void onClickNext() {
+        Domain domain = mDomainList.get(selectedTypePosition);
+        Intent intent = new Intent(mContext, MainActivity.class);
+        intent.putExtra(EXTRA_ID_DOMAIN, domain.getId());
+        mContext.startActivity(intent);
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: using ``Navigator`` to redirect screen

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java (hunk 5 above)
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: why don't you use ``Integer`` in here ?
It should be ``int``

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -29,4 +29,5 @@
     <string name="featured">Featured</string>
     <string name="new_">New</string>
     <string name="sale_off">Sale off</string>
+    <string name="cannot_load_the_data">Can\'t load the data</string>
```
Comment (daolq3012) [inline app/src/main/res/values/strings.xml]: ``Can\'t`` ?

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java (hunk 7 above)
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: Save ``idDomain`` to local in here

File: app/src/main/java/com/framgia/forder/data/source/local/DomainLocalDataSource.java [hunk 9]
```diff
@@ -1,13 +1,46 @@
...
+
+    private RealmApi mRealmApi;
+    private SharedPrefsApi mSharedPrefsApi;
+
+    public DomainLocalDataSource(@NonNull RealmApi realmApi,@NonNull SharedPrefsApi prefsApi) {
+        mRealmApi = realmApi;
+        mSharedPrefsApi = prefsApi;
+    }
+
+    public User getUser(){
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/DomainLocalDataSource.java]: change to 
```java
    public Observable<User> getUser(){
         return mRealmApi.realmGet(new Action2<Subscriber<? super User>, Realm>() {
             @Override
             public void call(Subscriber<? super User> subscriber, Realm realm) {
                 User user = realm.where(User.class).findFirst();
                 subscriber.onNext(user);
                 subscriber.onComplete();
             }
         });
         return user[0];
     }
```

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainActivity.java [hunk 10]
```diff
@@ -4,22 +4,38 @@
...
 
     private ChooseDomainContract.ViewModel mViewModel;
+    private Navigator mNavigator;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
 
-        mViewModel = new ChooseDomainViewModel();
+        mViewModel = new ChooseDomainViewModel(this, new Navigator(this));
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainActivity.java]: remove ``context`` params

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 11]
```diff
@@ -25,4 +64,42 @@ public void onStop() {
...
+        mAdapterDomain.notifyDataSetChanged();
+        for (Domain domain : domains) {
+            mAdapter.addAll(domain.getName());
+        }
+        mAdapter.notifyDataSetChanged();
+    }
+
+    @Override
+    public void onGetDomainError() {
+        Toast.makeText(mContext, R.string.cannot_load_the_data, Toast.LENGTH_SHORT).show();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: remove ``Toast`` in here
add ``// TODO show dialog error later``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 12]
```diff
@@ -25,4 +64,42 @@ public void onStop() {
...
+    }
+
+    public void onClickNext(View view) {
+        Domain domain = mDomainList.get(mSelectedTypePosition);
+        Bundle bundle=new Bundle();
+        bundle.putInt(EXTRA_ID_DOMAIN, domain.getId());
+        mNavigator.startActivity(MainActivity.class, bundle);
+        mPresenter.saveDomainId(domain.getId());
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: move to above 
``mNavigator.startActivity(MainActivity.class, bundle);``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 13]
```diff
@@ -1,14 +1,46 @@
...
 
-public class ChooseDomainViewModel implements ChooseDomainContract.ViewModel {
+public class ChooseDomainViewModel extends BaseObservable
+        implements ChooseDomainContract.ViewModel {
+
+    private static final String EXTRA_ID_DOMAIN = "EXTRA_ID_DOMAIN";
 
     private ChooseDomainContract.Presenter mPresenter;
+    private ArrayAdapter<String> mAdapter;
+    private List<String> mDomains;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: remove

Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: remove

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 14]
```diff
@@ -1,14 +1,46 @@
...
+    private ArrayAdapter<Domain> mAdapterDomain;
+    private List<Domain> mDomainList;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: mDomains

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 15]
```diff
@@ -25,4 +57,42 @@ public void onStop() {
...
+    public void onGetDomainSuccess(List<Domain> domains) {
+        if (domains == null) {
+            return;
+        }
+        mDomainList.clear();
+        mDomains.clear();
+        mDomainList.addAll(domains);
+        mAdapterDomain.notifyDataSetChanged();
+        for (Domain domain : domains) {
+            mAdapter.addAll(domain.getName());
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: Change to ``mAdapter.add(domain.getName());``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 16]
```diff
@@ -25,4 +57,42 @@ public void onStop() {
...
+
+    @Override
+    public void onGetDomainError() {
+        // TODO show dialog error later
+    }
+
+    public void onClickNext(View view) {
+        Domain domain = mDomainList.get(mSelectedTypePosition);
+        mPresenter.saveDomainId(domain.getId());
+        Bundle bundle = new Bundle();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: remove ``Bundle``

File: app/src/main/java/com/framgia/forder/data/model/User.java [hunk 17]
```diff
@@ -104,6 +104,10 @@
     @SerializedName("updated_at")
     @Expose
     private String updatedAt;
+    @SerializedName("authentication_token")
+    @Expose
+    private String token;
```
Comment (DoanVanToan) [inline app/src/main/java/com/framgia/forder/data/model/User.java]: pls check convention

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainPresenter.java [hunk 18]
```diff
@@ -9,16 +18,46 @@
...
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(new Action1<List<Domain>>() {
+                    @Override
+                    public void call(List<Domain> domains) {
+                        mViewModel.onGetDomainSuccess(domains);
+                    }
+                }, new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable thrxowable) {
+                        mViewModel.onGetDomainError();
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainPresenter.java]: change to ``mViewModel.onGetDomainError(throwable);``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainPresenter.java [hunk 19]
```diff
@@ -9,16 +18,46 @@
...
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(new Action1<List<Domain>>() {
+                    @Override
+                    public void call(List<Domain> domains) {
+                        mViewModel.onGetDomainSuccess(domains);
+                    }
+                }, new Action1<Throwable>() {
+                    @Override
+                    public void call(Throwable thrxowable) {
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainPresenter.java]: change ``thrxowable`` to ``throwable``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 20]
```diff
@@ -1,14 +1,40 @@
...
  */
 
-public class ChooseDomainViewModel implements ChooseDomainContract.ViewModel {
+public class ChooseDomainViewModel extends BaseObservable
+        implements ChooseDomainContract.ViewModel {
 
     private ChooseDomainContract.Presenter mPresenter;
+    private ArrayAdapter<String> mAdapter;
+    private ArrayAdapter<Domain> mAdapterDomain;
+    private List<Domain> mDomainList;
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: change to ``mListDomain`` or ``mDomains``

File: app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java [hunk 21]
```diff
@@ -1,14 +1,40 @@
...
+    public ChooseDomainViewModel(ArrayAdapter<String> adapter, ArrayAdapter<Domain> adapterDomain,
+            Navigator navigator) {
+        mDomainList = new ArrayList<>();
+        mNavigator = navigator;
+        mAdapter = adapter;
+        mAdapterDomain = adapterDomain;
+    }
 
-    public ChooseDomainViewModel() {
+    @Bindable
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/screen/chooseDomain/ChooseDomainViewModel.java]: remove ``@Bindable`` it unnecessary
only use `@Bindable`` when need notifyDataChange(BR...)

File: app/src/main/java/com/framgia/forder/data/source/local/sharedprf/SharedPrefsKey.java [hunk 22]
```diff
@@ -7,6 +7,9 @@
 public final class SharedPrefsKey {
     public static final String KEY_USER = "KEY_USER";
 
+    public static final String KEY_DOMAIN_ID = "KEY_DOMAIN_ID";
+    public static final String KEY_USER = "KEY_USER";
```
Comment (daolq3012) [inline app/src/main/java/com/framgia/forder/data/source/local/sharedprf/SharedPrefsKey.java]: remove

**Pull request merged**